packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to either define the field to sort by, or provide a custom sort routine. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```
+<nimble-table>
+    <nimble-table-column-text valueKey="name"></nimble-table-column-text>
+    <nimble-table-column-icon valueKey="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation.
+
+The table API to support this could look like the following:
+
+```TS
+public Table extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: ITableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
```

Comment:
Can you clarify what the purpose of `isTableColumn()` is? If the `slottedColumns` are all `ITableColumn` then when would that method return `false`?

Reply:
Maybe this is too much in the implementation details but we can use the filter option on the slotted directive for this: https://www.fast.design/docs/fast-element/using-directives#the-slotted-directive

If we use a concrete TableColumn class then we can do `instanceof` checks.

For an interface we can't rely on `instanceof` but maybe instead we can expose a TableColumn Symbol that columns set as a known property to indicate they fulfill the interface (like [iterators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator#user-defined_iterables)). We can enforce in the type that they put a required symbol, [this playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAKgQwEYBsCmBhEKCuBbMALhhzAEsBHHNGCATzyWxgF4YBlBplACgEoA3ACgodAA41EqTNnxhWMURJAAzeMnRZcBYULJgoaAE4qEwGgEkpm2QRgBvITGcwjaBABNwKOjADa1jLaYAC6xFBG1MIuru5eYD6KGkFyxIFacsIAvkJCwCgIEBAwAIIA7mgQIHgpdmR4Yug1BsVWyRl2jjFiOKhkwLGe3r5Q7bZEiuJoqurSHfJs6ePRLj19A25DCb4BY8EhChFRQjlAA) shows two ways we could approach that.

Reply:
Yes, this is perhaps a bit too much implementation detail, but this code was simply meant to reflect the possible need to differentiate between other content that was put in the default slot of the table.

Reply:
> this code was simply meant to reflect the possible need to differentiate between other content that was put in the default slot of the table.

I'd expect any alternate content to be assigned to it's own slot like the action menu. We can do this for validation / filtering user input (and how validation is done is implementation detail) but I wouldn't expect the public api to consist of mixing things besides column configuration elements (ie column config and column group config).

So just to confirm, example looks good as long as default slot is just for column config elements